File: 22.generate_parentheses.js
// https://leetcode.com/problems/generate-parentheses/description/

// Input: n = 3
// Output: ["((()))","(()())","(())()","()(())","()()()"]

/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
  const r = []
  function backtrack(path, lc, rc) {
    // base cases
    if (rc > lc) {
      return
    }
    if (path.length === n * 2) {
      r.push(path.join(""))
      return
    }

    // go left
    if (lc < n) {
      path.push("(")
      backtrack(path, lc + 1, rc)
      path.pop()
    }

    // go right
    if (rc < n) {
      path.push(")")
      backtrack(path, lc, rc + 1)
      path.pop()
    }
  }

  backtrack([], 0, 0)
  return r
}

console.log(generateParenthesis(3))
// console.log(generateParenthesis(6))

File: 22.generate_parentheses.py
# https://leetcode.com/problems/generate-parentheses/description/


from typing import List

# Input: n = 3
# Output: ["((()))","(()())","(())()","()(())","()()()"]

# Topics
# String
# Dynamic Programming
# Backtracking


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        r = []

        # binary decision tree backtracking
        # lc: "(" count
        # rc: ")" count
        def backtrack(path: List[str], lc, rc):
            # base cases
            if rc > lc:
                return
            if len(path) == n * 2:
                r.append("".join(path))
                return

            # go left
            if lc < n:
                path.append("(")
                backtrack(path, lc + 1, rc)
                path.pop()

            # go right
            if rc < n:
                path.append(")")
                backtrack(path, lc, rc + 1)
                path.pop()

            return

        backtrack([], 0, 0)
        return r


s = Solution()
# print(s.generateParenthesis(1))
print(s.generateParenthesis(3))
# print(s.generateParenthesis(6))
